tests/rate/endpoint_rate_limit_test.go

```diff
@@ -0,0 +1,211 @@
+package rate_test
+
+import (
+	"net/http"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/assert"
+
+	. "github.com/TykTechnologies/tyk/gateway"
+	"github.com/TykTechnologies/tyk/header"
+	"github.com/TykTechnologies/tyk/test"
+	"github.com/TykTechnologies/tyk/user"
+)
+
+func rlTestRunnerProvider(t *testing.T, hashKey bool, hashAlgo string, limiter string) *Test {
+	t.Helper()
+	ts := StartTest(nil)
+	defer ts.Close()
+
+	globalConf := ts.Gw.GetConfig()
+
+	globalConf.HashKeys = hashKey
+	globalConf.HashKeyFunction = hashAlgo
+
+	switch limiter {
+	case "Redis":
+		globalConf.RateLimit.EnableRedisRollingLimiter = true
+	case "Sentinel":
+		globalConf.RateLimit.EnableSentinelRateLimiter = true
+	case "DRL":
+		globalConf.RateLimit.DRLEnableSentinelRateLimiter = true
+	case "NonTransactional":
+		globalConf.RateLimit.EnableNonTransactionalRateLimiter = true
+	default:
+		t.Fatal("There is no such a rate limiter:", limiter)
+	}
+
+	ts.Gw.SetConfig(globalConf)
+
+	ok := ts.Gw.GlobalSessionManager.Store().DeleteAllKeys()
+	assert.True(t, ok)
+
+	return ts
+}
+
+func endpointRateLimitTestHelper(t *testing.T, limiter string, beforeFn func()) {
```

Comment:
try to follow tests/regression structure on internals, put these in rate.go more files is ok